src/tbl_subassign_col.c

```diff
@@ -0,0 +1,131 @@
+#include "tibble.h"
+
+R_xlen_t fast_nrow(SEXP x);
+
+// Performs a variant of x[j] <- value, growing x as necessary
+// NULL values are supported
+// First, creates the storage for the resulting x
+// Then, populates that storage by copying the correct values exactly once
+//
+// Requires j to be sorted on input
+SEXP tbl_subassign_col(SEXP x, SEXP j_, SEXP value) {
+  // Naming conventions:
+  // n_x: length of x
+  // x_idx: index into vector x
+  // x_names: names of x
+  // n_old: number of elements with some property
+  // j_max: maximum value of j
+  // j_/j: SEXP and native version of bare vector
+  // xo: output vector
+  R_xlen_t nrow = fast_nrow(x);
+  R_xlen_t n_x = Rf_xlength(x);
+
+  // Compute j from j_, converting 1-based j_ to zero-based j
+  R_xlen_t n_j = Rf_xlength(j_);
+  R_xlen_t* j = (R_xlen_t*)R_alloc(n_j + 1, sizeof(R_xlen_t));
```

Comment:
Could you have used `vec_as_location()` upstream to avoid having to deal with double vectors? I imagine tibble doesn't need support for long vectors in the column dimension.